base/abstractarray.jl

```diff
@@ -422,6 +422,21 @@ function _maxlength(A, B, C...)
     max(length(A), _maxlength(B, C...))
 end
 
+# eachindex(A, d) and eachindex(A, (:, j))
+eachindex(A::AbstractArray, d::Integer) = 1:size(A, d)
```

Comment:
How about `dims` as a keyword argument? Otherwise I don't think trying to cram too many things at once into the same `eachindex` god-function is necessarily a good idea.